### PR TITLE
(655c) Create course participation from select programme page

### PR DESCRIPTION
**Pull request by jsrobertson**
## Context

https://trello.com/c/i2gzZOoW/655-add-page-for-selecting-course-name-ui

## Changes in this PR
Note: Branched off #218 so contains changes to courseClient and courseService until that has been merged. 

Updates the select a programme form page to `POST` to `/course-participations` and redirects to the yet to be created details page. 

Also displays validation for either an unselected programme or missing text in the conditionally displayed Other text input.

## Screenshots of UI changes

localhost_3000_referrals_c11fab18-dc8d-420c-9c82-d0edd373732d_programme-history_new (7)

localhost_3000_referrals_c11fab18-dc8d-420c-9c82-d0edd373732d_programme-history_new (8)

localhost_3000_referrals_c11fab18-dc8d-420c-9c82-d0edd373732d_programme-history_new (9)

## Release checklist

[Release process documentation](../doc/how-to/perform-a-release.md)

As part of our continuous deployment strategy we must ensure that this work is
ready to be released once merged.

### Pre-merge

- [x] There are changes required to the Accredited Programmes API for this change to work...
  - [ ] ... and they have been released to production already

### Post-merge

- [x] This adds, extends or meaningfully modifies a feature...
  - [ ] ... and I have written or updated an end-to-end test for the happy path in the [Accredited Programmes E2E repo](https://github.com/ministryofjustice/hmpps-accredited-programmes-e2e)
- [ ] This makes new expectations of the API...
  - [ ] ... and I have notified the API developer(s) of changes to the contract tests (Pact), or the API is already compliant
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-preprod-environment) release to preprod
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-production-environment) release to prod

### CODE REVIEW

File: server/views/referrals/courseParticipations/new.njk [hunk 1]
```diff
@@ -47,8 +56,10 @@
             text: "Other",
             conditional: {
               html: otherCourseHtml
-            }
-          })
+            },
+            checked: errors.messages.otherCourseName
```
Comment (yndajas) [inline server/views/referrals/courseParticipations/new.njk]: Nice

File: integration_tests/pages/page.ts [hunk 2]
```diff
@@ -138,6 +142,21 @@ export default abstract class Page {
     })
   }
 
+  shouldContainRadioButtons(options: Array<{ label: string; value: string }>): void {
```
Comment (yndajas) [inline integration_tests/pages/page.ts:145]: Are the contents of the array here `GovukFrontendRadiosItem`s (possibly `WithLabel` to assert that they aren't `undefined`)? If that's the right type, this might be more of a `shouldContainRadioItems` method with an `items` argument

File: integration_tests/pages/page.ts [hunk 3]
```diff
@@ -138,6 +142,21 @@ export default abstract class Page {
...
+    cy.get('.govuk-radios__item').each((radioButtonElement, radioButtonElementIndex) => {
+      const option = options[radioButtonElementIndex]
```
Comment (yndajas) [inline integration_tests/pages/page.ts:147]: Is it worth flipping this so that we can use the `.eq` method you used before here? E.g. `options(or items).forEach((option, optionIndex) => cy.get('.govuk-radios__item').eq(optionIndex)...`?

File: integration_tests/pages/page.ts (hunk 2 above)
Comment (yndajas) [inline integration_tests/pages/page.ts:145]: Would this method work with multiple radio input questions on a page? (The details page will have this) Does the containing `.govuk-radios` element need passing in to constrain it?

Perhaps that iteration can come when adding the details page

File: integration_tests/pages/refer/selectProgramme.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
+import Page from '../page'
+import type { Course } from '@accredited-programmes/models'
+
+export default class SelectProgrammePage extends Page {
+  courses: Array<Course>
+
+  constructor(args: { courses: Array<Course> }) {
+    // Conditional radio buttons add an additional `aria-expanded` field,
+    // so ignore that rule on this page
+    super('Add Accredited Programme history', { accessibilityRules: { 'aria-allowed-attr': { enabled: false } } })
```
Comment (yndajas) [inline integration_tests/pages/refer/selectProgramme.ts:10]: Is it worth making the exception more specific, e.g. to a particular selector like here: https://github.com/ministryofjustice/hmpps-accredited-programmes-ui/blob/main/integration_tests/support/commands.ts#L29. We still want to check this rule outside the conditional input, right?

File: integration_tests/e2e/refer.cy.ts [hunk 5]
```diff
@@ -344,6 +346,80 @@ context('Refer', () => {
     })
   })
 
+  describe('When selecting a programme', () => {
```
Comment (yndajas) [inline integration_tests/e2e/refer.cy.ts:349]: Might be worth nesting this in `'When adding programme history'` for context

File: integration_tests/e2e/refer.cy.ts [hunk 6]
```diff
@@ -344,6 +346,80 @@ context('Refer', () => {
...
+      cy.task('stubReferral', referral)
+
+      cy.signIn()
+
+      cy.visit(path)
+    })
+
+    it('shows the select programme form page', () => {
+      const selectProgrammePage = Page.verifyOnPage(SelectProgrammePage, { courses })
+      selectProgrammePage.shouldContainCourseOptions()
```
Comment (yndajas) [inline integration_tests/e2e/refer.cy.ts:374]: We should have the standard Refer assertions here too:

```ts
    confirmOasysPage.shouldHavePersonDetails(person)
    confirmOasysPage.shouldContainNavigation(path)
    confirmOasysPage.shouldContainBackLink('#')
```

File: integration_tests/e2e/refer.cy.ts [hunk 7]
```diff
@@ -344,6 +346,80 @@ context('Refer', () => {
...
+      selectProgrammePage.shouldNotDisplayOtherCourseInput()
+      selectProgrammePage.shouldDisplayOtherCourseInput()
+    })
```
Comment (yndajas) [inline integration_tests/e2e/refer.cy.ts:377]: And an assertion about the continue button

File: integration_tests/e2e/refer.cy.ts [hunk 8]
```diff
@@ -344,6 +346,80 @@ context('Refer', () => {
...
+
+    it('creates a course participation and redirects to the details page', () => {
+      const courseParticipation = courseParticipationFactory.withCourseId().build({ courseId: courses[0].id })
```
Comment (yndajas) [inline integration_tests/e2e/refer.cy.ts:380]: Should this have the prison number too?

File: integration_tests/mockApis/courseParticipations.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
+import type { SuperAgentRequest } from 'superagent'
+
+import { apiPaths } from '../../server/paths'
+import { stubFor } from '../../wiremock'
+import type { CourseParticipation, Person } from '@accredited-programmes/models'
+
+export default {
+  stubCreateParticipation: (args: {
+    courseParticipation: CourseParticipation
+    prisonNumber: Person['prisonNumber']
```
Comment (yndajas) [inline integration_tests/mockApis/courseParticipations.ts:10]: The `prisonNumber` isn't user here (and could be taken from the course participation if needed)

**Review by yndajas — COMMENTED** [review]
A few tidy up suggestions, but generally looks good

File: server/data/courseClient.test.ts [hunk 10]
```diff
@@ -67,6 +68,68 @@ pactWith({ consumer: 'Accredited Programmes UI', provider: 'Accredited Programme
     })
   })
 
+  describe('createParticipation', () => {
+    describe('when `courseId` is provided', () => {
+      const courseParticipation = courseParticipationFactory.withCourseId().build()
+      const { id: courseId, prisonNumber } = courseParticipation
```
[the remark this answers is not in the document]
Comment (jsrobertson) [reply server/data/courseClient.test.ts:74]: Good spot, it is meant to be just `courseId`, not renaming `id` to `courseId` as I've done here.

File: integration_tests/mockApis/courseParticipations.ts (hunk 9 above)
Comment (jsrobertson) [reply integration_tests/mockApis/courseParticipations.ts:10]: Good spot, intended to use it but then turned out I didn't need it.

File: integration_tests/e2e/refer.cy.ts [hunk 11]
```diff
@@ -344,6 +346,80 @@ context('Refer', () => {
...
+    beforeEach(() => {
+      cy.task('stubCourses', courses)
+      cy.task('stubReferral', referral)
+
+      cy.signIn()
+
+      cy.visit(path)
+    })
+
+    it('shows the select programme form page', () => {
```
Comment (yndajas) [inline integration_tests/e2e/refer.cy.ts:372]: We should also have a corresponding test in `e2eReferDisabled`

File: server/data/courseClient.ts [hunk 12]
```diff
@@ -15,6 +15,17 @@ export default class CourseClient {
...
 
+  async createParticipation(
+    prisonNumber: CourseParticipation['prisonNumber'],
+    courseId?: CourseParticipation['id'],
+    otherCourseName?: string,
+  ): Promise<CourseParticipation> {
+    return this.restClient.post({
+      data: { courseId, otherCourseName, prisonNumber },
+      path: apiPaths.participations.create({}),
+    }) as Promise<CourseParticipation>
```
[the remark this answers is not in the document]
Comment (yndajas) [reply server/data/courseClient.ts:26]: (Realised you said as much in the previous PR - looks like I reviewed a couple of commits twice!)

File: server/data/courseClient.ts [hunk 13]
```diff
@@ -15,6 +15,17 @@ export default class CourseClient {
     return (await this.restClient.get({ path: apiPaths.courses.index({}) })) as Array<Course>
   }
 
+  async createParticipation(
+    prisonNumber: CourseParticipation['prisonNumber'],
+    courseId?: CourseParticipation['id'],
+    otherCourseName?: string,
```
Comment (yndajas) [inline server/data/courseClient.ts:21]: The type of `otherCourseName` here can also come from `CourseParticipation`

File: server/controllers/refer/courseParticipationsController.ts [hunk 14]
```diff
@@ -1,20 +1,47 @@
...
     return async (req, res) => {
       TypeUtils.assertHasUser(req)
 
       const courses = await this.courseService.getCourses(req.user.token)
+      const referral = await this.referralService.getReferral(req.user.token, req.params.referralId)
 
       res.render('referrals/courseParticipations/new', {
         courseRadioOptions: CourseUtils.courseRadioOptions(courses),
         pageHeading: 'Add Accredited Programme history',
+        referral,
```
[the remark this answers is not in the document]
Comment (jsrobertson) [reply server/controllers/refer/courseParticipationsController.ts]: Yeah makes sense to just have `referralId`, also moved to appropriate commit.

File: server/data/courseClient.test.ts [hunk 15]
```diff
@@ -67,6 +68,68 @@ pactWith({ consumer: 'Accredited Programmes UI', provider: 'Accredited Programme
...
+      beforeEach(() => {
+        provider.addInteraction({
+          state: `A person exists with prison number ${prisonNumber}`,
+          uponReceiving: `A request to create a course participation for person "${prisonNumber}"`,
+          willRespondWith: {
+            body: Matchers.like(courseParticipation),
+            status: 201,
+          },
+          withRequest: {
+            body: { otherCourseName, prisonNumber } as CourseParticipation,
```
[the remark this answers is not in the document]
Comment (jsrobertson) [reply server/data/courseClient.test.ts:115]: Also needed to cast it in the previous test once I had changed `id: courseId` to just `courseId`.

File: server/services/courseService.ts [hunk 16]
```diff
@@ -4,6 +4,21 @@ import type { Course, CourseOffering, CourseParticipation, Person } from '@accre
...
+    token: Express.User['token'],
+    prisonNumber: CourseParticipation['prisonNumber'],
+    courseId?: CourseParticipation['id'],
+    otherCourseName?: CourseParticipation['otherCourseName'],
+  ): Promise<CourseParticipation> {
+    const courseClient = this.courseClientBuilder(token)
+
+    if (courseId) {
+      return courseClient.createParticipation(prisonNumber, courseId)
+    }
```
[the remark this answers is not in the document]
Comment (jsrobertson) [reply server/services/courseService.ts:17]: Did do that to begin with so not sure why I changed to this way.  Have reverted and updated in the other branch too.

File: server/data/courseClient.test.ts [hunk 17]
```diff
@@ -67,6 +68,68 @@ pactWith({ consumer: 'Accredited Programmes UI', provider: 'Accredited Programme
...
+            method: 'POST',
+            path: apiPaths.participations.create({}),
+          },
+        })
+      })
+
+      it('creates a course participation using the `courseId` value', async () => {
+        const result = await courseClient.createParticipation(prisonNumber, courseId)
+
+        expect(result).toEqual(courseParticipation)
```
[the remark this answers is not in the document]
Comment (jsrobertson) [reply server/data/courseClient.test.ts:98]: Agreed, a started trait would probably make more sense.  Will address in another PR though.

File: integration_tests/pages/page.ts (hunk 2 above)
Comment (jsrobertson) [reply integration_tests/pages/page.ts:146]: Will address that separately.

File: integration_tests/pages/refer/selectProgramme.ts (hunk 4 above)
Comment (jsrobertson) [reply integration_tests/pages/refer/selectProgramme.ts:10]: Yeah possibly, might need to make a change to `Cypress.Commands.add('checkAccessibility`.  Will investigate separately.

File: integration_tests/e2e/refer.cy.ts [hunk 18]
```diff
@@ -349,6 +350,92 @@ context('Refer', () => {
...
+        cy.signIn()
+
+        cy.visit(path)
+      })
+
+      it('shows the select programme form page', () => {
+        const selectProgrammePage = Page.verifyOnPage(SelectProgrammePage, { courses })
+        selectProgrammePage.shouldHavePersonDetails(person)
+        selectProgrammePage.shouldContainNavigation(path)
+        selectProgrammePage.shouldContainBackLink('#')
```
Comment (Gweaton) [inline integration_tests/e2e/refer.cy.ts:382]: Now that we have the programme history index page, shall we put these back links in? Happy to do that in a separate PR if you'd rather but shouldn't be a big change.

**Review by Gweaton — APPROVED** [review]
Nice work on this, one small thing that might be good to add in

File: integration_tests/e2e/refer.cy.ts (hunk 18 above)
Comment (jsrobertson) [reply integration_tests/e2e/refer.cy.ts:382]: Aye good shout
